code/cdw.py

```diff
@@ -0,0 +1,181 @@
+    mask : Nifti1Image
+        A 3D mask to restrict the analysis with.
+    rad : int
+        Radius of searchlight (voxels). This currently defines a square, not
+        a sphere -- subject to revision at a later point.
+    conds : dict
+        Dictionary with condition names as keys and arrays of MRI frame
+        indices as values.
+
+    Returns
+    -------
+    rsa_brain : numpy array
+        RSA-scores in brain space.
+    """
+    
+    rsa_brain = np.zeros(mask.shape)
+    
+    # Reorder bold data into "condition order"  & drop extra frames
+    print('Reordering data and dropping unnecessary MRI frames...')
+    bold_conds = np.concatenate(list(time2conds(bold_orig, conds)), axis=-1)
+    print('Done. Calculating RSA...')
+    
+    rdm_model = create_modelRDM(conds)
+
+    inds = np.where(mask.get_fdata() == 1)
+    print(f'Total number of voxels: {len(inds[0])}')
+    
+    # For each voxel in mask:
+    i = 0
+    for ind in zip(inds[0], inds[1], inds[2]):
+        
+        print(f'Processing voxel {ind} -- {i}')
+        
+        # Calc rdm in searchlight
+        rdm_bold = calculate_boldRDM(bold_conds, ind, rad)
+        
+        # Calc rsa and save it
+        rsa_brain[ind] = calculate_RSA(rdm_bold, rdm_model)
+        
+        i += 1
+        if i > 10:
+            break
+        
+    # Wrap up as nifti
+    rsa_brain_nii = nib.Nifti1Image(rsa_brain, bold_orig.affine, bold_orig.header)
+    
+    # return rsa brain
+    print('Done.')
+    return rsa_brain_nii
+
+# =============================================================================
+#                        INTERNAL HELPER FUNCTIONS
+# =============================================================================
+
+def create_modelRDM(conds):
+    # makes the model rdm, given conds
+    
+    # Get framecounts
+    lens = [len(conds[k]) for k in conds.keys()]
+    start_inds = np.cumsum(lens)
+    start_inds = np.insert(start_inds,0,0)
+    
+    # Fill matrix according to framecounts
+    rdm_model = np.ones((sum(lens),sum(lens)))
+    for i in range(len(conds)):
+        start = start_inds[i]
+        stop = start_inds[i+1]
+        rdm_model[start:stop, start:stop] = 0 
+    
+    return rdm_model
+
+def calculate_boldRDM(bold, ind, rad):
+    # calc the rdm in given searchlight beam
+    
+    # Get voxels inside radius (square for now)
+    voxels = bold[ind[0]-rad: ind[0]+rad+1,
+                   ind[1]-rad: ind[1]+rad+1,
+                   ind[2]-rad: ind[2]+rad+1, :]
+    
+    # Flatten in all but last dimension (MRI frames)
+    n_frames = voxels.shape[-1]
+    voxels = np.vstack([ voxels[:,:,:,i].flatten() for i in range(n_frames) ])
+    
+    # Calculate dissimilarity (1 - r) for each frame pair
+    rdm_bold = np.ones((n_frames,n_frames))
+    for i in range(n_frames):
+        for j in range(n_frames):
+            (r,p) = stats.pearsonr(voxels[i,:], voxels[j,:]) # mri frames are rows
+            rdm_bold[i,j] = 1 - r
```

Comment:
```suggestion
    from scipy.spatial.distance import pdist, squareform
    rdm_bold = squareform(pdist(voxels, metric='correlation'))
```

Reply:
Ahh I already made a bit if a speed-up for this (true, this version was _incredibly_ slow), but apparently forgot to push it! Anyway, I'll check these functions as well, this is way more concise.